refactors the internal typing and modeling class

privateip opened the pull request:
This is a fairly significant refactor of the type and model system used
network-runner.  It decouples the type and resource objects into
different packages and substaintially streamlines the type system.  Test
cases have been added to validate the new modules.

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
refactors the interal typing and modeling class

This is a fairly significant refactor of the type and model system used
network-runner.  It decouples the type and resource objects into
different packages and substaintially streamlines the type system.  Test
cases have been added to validate the new modules.

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
remove unused kwarg in Atribute()

Commit:
removes restriction for using underscore in attribute definition

This commit removes the need to define attributes with underscores and
now attributes can be propertly named.

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
fixes issue with task serialization / deserialization

This fixes a problem with how the Task resource is serialized and
deserialized.  This change also corrects the Task reosurce to use action
and args instead of trying to generate an action key for the play.

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

privateip commented:
recheck

Commit:
fix bug introduced into inventory object for null variable data

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
remove accidentially added context module

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

radez commented on network_runner/resources/inventory.py:
```diff
@@ -0,0 +1,116 @@
+        serialize_when=SERIALIZE_WHEN_PRESENT
+    )
+
+    ansible_password = Attribute(
```
We had trouble passing ansible_password instead of ansible_ssh_pass. The ML2 driver assumes ssh_pass.
Could we keep it ssh_pass or handle both we don't break the ml2 driver? We'll probably need to test the transition outside of this patch to make sure things don't break.

privateip replied:
Just pushed a change that allows attributes to define one or more aliases.  Setting one will set all.  Let me know if that addresses your concern.

radez commented on tests/unit/test_resources.py:
```diff
@@ -56,19 +56,19 @@ def test_host_attrs(self):
         self.assertEqual(host.ansible_host, 'testhost')
         self.assertEqual(host.ansible_user, 'testuser')
-        self.assertEqual(host.ansible_ssh_pass, 'testpass')
+        self.assertEqual(host.ansible_password, 'testpass')
```
see the comment about ssh_pass on resources.inventory

radez commented on tests/unit/test_resources.py:
```diff
@@ -56,19 +56,19 @@ def test_host_attrs(self):
                     ansible_host='testhost',
                     ansible_user='testuser',
-                    ansible_ssh_pass='testpass',
+                    ansible_password='testpass',
```
see the comment about ssh_pass on resources.inventory

radez reviewed (commented):
Looks good other than the password handling syntax

Commit:
allows attributes to have one or more aliases

This change allows attributes to define a list of alias names.  When
configured, setting any one attribute will update all aliases as well.

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
fixes issue with assign arbitrary vars

This commit fixes an situation around passing arbitirary kwargs when
creating a Host or Child instance.  Once applied, arbitrary kwargs will
be added to the instance vars

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
fixes issue with bool not set when passed as kwargs

Signed-off-by: Peter Sprygada <psprygad@redhat.com>

Commit:
increases unit test coverage and removes unnecessary methods

Signed-off-by: Peter Sprygada <psprygad@redhat.com>
